src/ExperimentalFunds.mo

```diff
@@ -0,0 +1,54 @@
+/// Funds and units
+///
+/// Provides imperative operations for observing funds, transferring funds and
+/// observing refunds of various units.
+/// The two currently support units are `#cycle` and `#icpt`, measuring
+/// computational cost and Internet Computer tokens, respectively.
+
+/// This low-level API is experimental and likely to change, including the
+/// addition of other units and alternative syntactic support for
+/// manipulating funds.
+///
+/// Note: since unit `#cycle` is used to measure computation, the value of
+/// `balance(#cycle)` generally decreases from one call to the next.
+
+import Prim "mo:prim";
+module {
+
+  /// Units for funds: `{#cycle; #icpt}`.
+  public type Unit = Prim.Unit;
+
+  /// Returns the actor's current balance of unit `u` as `amount`.
+  public let balance : (u : Unit) -> (amount : Nat64) = Prim.fundsBalance;
+
+  /// Given `u`, returns the currently available `amount` of unit `u`.
+  /// The amount available is the amount received in the current message,
+  /// minus the cumulative amount `accept`ed by this message.
+  /// On return to the sender, any remaining available amount is automatically
+  /// refunded to the sender.
+  public let available : (u : Unit) -> (amount : Nat64) = Prim.fundsAvailable;
+
+  /// Transfers `amount` from `available(u)` to `balance(u)`,
+  /// trapping on underflow.
+  public let accept : (u : Unit, amount : Nat64) -> () = Prim.fundsAccept;
+
+  /// Indicates additional `amount` of unit `u` to be transferred in
+  /// the next message, i.e. evaluation of a shared function call or
+  /// async expression.
+  /// On message send, but not before, the total amount of units `add`ed since
+  /// the last send is deducted from `balance(u)`.
```

Comment:
```suggestion
  /// Upon the call, but not before, the total amount of units `add`ed since
  /// the last call is deducted from `balance(u)`.
```

Reply:
Yeah I was deliberately using send rather than call because this applies to shared functions calls and async blocks. But I can change to (shared) call if you like